gradio/queueing.py

```diff
@@ -345,81 +368,70 @@ async def clean_events(
-        estimation = self.get_estimation()
-        # Send all messages concurrently
-        await asyncio.gather(
-            *[
-                self.send_estimation(event, estimation, rank)
-                for rank, event in enumerate(self.event_queue)
-            ]
-        )
-
-    async def send_estimation(
-        self, event: Event, estimation: Estimation, rank: int
-    ) -> Estimation:
-        """
-        Send estimation about ETA to the client.
+            if len(self) > 0:
+                for concurrency_id in self.event_queue_per_concurrency_id:
+                    self.broadcast_estimations(concurrency_id)
 
-        Parameters:
-            event:
-            estimation:
-            rank:
-        """
-        estimation.rank = rank
-
-        if self.avg_concurrent_process_time is not None:
-            estimation.rank_eta = (
-                estimation.rank * self.avg_concurrent_process_time
-                + self.avg_process_time
+    def broadcast_estimations(
+        self, concurrency_id: str, after: int | None = None
+    ) -> None:
+        wait_so_far = 0
+        event_queue = self.event_queue_per_concurrency_id[concurrency_id]
+        for rank, event in enumerate(event_queue.queue):
+            process_time_for_fn = (
+                self.process_time_per_fn_index[event.fn_index].avg_time
+                if event.fn_index in self.process_time_per_fn_index
+                else None
             )
-            if None not in self.active_jobs:
-                # Add estimated amount of time for a thread to get empty
-                estimation.rank_eta += self.avg_concurrent_process_time
-        self.send_message(event, ServerMessage.estimation, estimation.model_dump())
-        return estimation
-
-    def update_estimation(self, duration: float) -> None:
-        """
-        Update estimation by last x element's average duration.
-
-        Parameters:
-            duration:
-        """
-        self.duration_history_total += duration
-        self.duration_history_count += 1
-        self.avg_process_time = (
-            self.duration_history_total / self.duration_history_count
-        )
-        self.avg_concurrent_process_time = self.avg_process_time / min(
-            self.max_thread_count, self.duration_history_count
-        )
-        self.queue_duration = self.avg_concurrent_process_time * len(self.event_queue)
+            rank_eta = (
+                process_time_for_fn + wait_so_far
+                if process_time_for_fn is not None and wait_so_far is not None
+                else None
+            )
+            if (
+                rank_eta is not None
+                and event_queue.current_concurrency == event_queue.concurrency_limit
+                and process_time_for_fn is not None
+                and event_queue.concurrency_limit is not None
+            ):
+                rank_eta += process_time_for_fn / event_queue.concurrency_limit / 2
```

Comment:
What is this doing?

Reply:
If there are no available free workers due to hitting the concurrency limit, the expected amount of time we have to wait for a worker to free up is the expected runtime of the function divided by the concurrency limit divided by 2 (we divide by two assuming that any worker is half way through running the process)